packages/@aws-cdk/pipelines/lib/stage.ts

```diff
@@ -76,14 +81,79 @@ export class CdkStage extends CoreConstruct {
    */
   public addApplication(appStage: Stage, options: AddStageOptions = {}) {
     const asm = appStage.synth({ validateOnSynthesis: true });
-    const extraRunOrderSpace = options.extraRunOrderSpace ?? 0;
+    const extraRunOrderSpace = options.extraRunOrderSpace ?? 0 + (options.securityCheck ? 2 : 0);
 
     if (asm.stacks.length === 0) {
       // If we don't check here, a more puzzling "stage contains no actions"
       // error will be thrown come deployment time.
       throw new Error(`The given Stage construct ('${appStage.node.path}') should contain at least one Stack`);
     }
 
+    if (options.securityCheck) {
+      const preApproveLambda = new lambda.Function(this, `${this.stageName}AutoApprove`, {
+        functionName: 'AutoApproveLambda',
+        handler: 'index.handler',
+        runtime: lambda.Runtime.PYTHON_3_8,
+        code: lambda.Code.fromAsset(resolve(__dirname, 'lambda')),
+      });
+
+      const invokeLambda =
+        'aws lambda invoke' +
+        ' --function-name AutoApproveLambda' +
+        ' --invocation-type Event' +
+        ' --payload "$payload"' +
+        ' lambda.out';
+
+      const build = new codebuild.Project(this, 'CDKSecurityCheck', {
+        buildSpec: codebuild.BuildSpec.fromObject({
+          version: 0.2,
+          phases: {
+            build: {
+              commands: [
+                'npm install -g aws-cdk',
+                // $CODEBUILD_INITIATOR will always be Code Pipeline and in the form of:
+                // "codepipeline/example-pipeline-name-Xxx"
+                // cut splits the string based on the delimiter "/" and returns the 2nd field
+                'payload="$(jq -n --arg pipelineName "$(echo $CODEBUILD_INITIATOR | cut -d "/" -f 2 )" --arg stageName "$STAGE_NAME" \'{ PipelineName: $pipelineName, StageName: $stageName }\' )"',
+                'cdk diff -a . --security-only --fail',
+                `$? || ${invokeLambda}`,
+              ],
+            },
+          },
+        }),
+      });
+
+      preApproveLambda.grantInvoke(build);
+
+      build.addToRolePolicy(new iam.PolicyStatement({
+        actions: ['cloudformation:DescribeStacks', 'cloudformation:GetTemplate'],
+        resources: ['*'], // this is needed to check the status the stacks when doing `cdk diff`
```

Comment:
I think what's necessary is that this stage is able to assume the `deploy-role` for this account.

Reply:
Why does the `deploy-role` matter? The only thing diff requires is to get the template and describe the stack no?

Reply:
It matters if the stack is in a different account or region.

`cdk diff` will do this automatically if given the opportunity (the fact that it works without assuming the role is for backwards compatibility--if you look at the logs you will most likely see it complain in there)